xds/src/main/java/io/grpc/xds/XdsLbState.java

```diff
@@ -149,30 +433,252 @@ public boolean hasNonDropBackends() {
+        }
+      }
     }
 
     @Override
     public void shutdown() {
-      // TODO: impl
+      shutdown = true;
+      for (IntraLocalitySubchannelStore subchannels : localityStore.values()) {
+        subchannels.shutdown();
+      }
+      localityStore = ImmutableMap.of();
+    }
+
+    // This is triggered by EDS response.
+    @Override
+    public void updateLocalityStore(Map<Locality, LocalityInfo> localityInfoMap) {
+      if (shutdown) {
+        return;
+      }
+
+      Set<Locality> oldLocalities = localityStore.keySet();
+      Set<Locality> newLocalities = localityInfoMap.keySet();
+
+      for (Locality oldLocality : oldLocalities) {
+        if (!newLocalities.contains(oldLocality)) {
+          // No graceful transition until a high-level lb graceful transition design is available.
+          localityStore.get(oldLocality).shutdown();
+          localityStore.remove(oldLocality);
+          if (localityStore.isEmpty()) {
+            // down-size the map
+            localityStore = new HashMap<>();
+          }
+        }
+      }
+
+      ConnectivityState newState = null;
+      List<LocalityState> localityStates = new ArrayList<>(newLocalities.size());
+      Map<Locality, SubchannelPicker> childPickers = new HashMap<>();
+      for (Locality newLocality : newLocalities) {
+
+        // Assuming standard mode only (EDS response with a list of endpoints) for now
+        List<EquivalentAddressGroup> newEags = localityInfoMap.get(newLocality).eags;
+        IntraLocalitySubchannelStore intraLocalitySubchannelStore;
+        if (oldLocalities.contains(newLocality)) {
+          intraLocalitySubchannelStore = localityStore.get(newLocality);
+        } else {
+          intraLocalitySubchannelStore =
+              new IntraLocalitySubchannelStore(newLocality, loadBalancerProvider);
+          localityStore.put(newLocality, intraLocalitySubchannelStore);
+        }
+        // TODO: put endPointWeights into attributes for WRR.
+        intraLocalitySubchannelStore.childBalancer
+            .handleResolvedAddresses(
+                ResolvedAddresses.newBuilder().setServers(newEags).build());
+
+        LocalityState localityState =
+            new LocalityState(
+                newLocality,
+                localityInfoMap.get(newLocality).localityWeight,
+                intraLocalitySubchannelStore.currentChildState);
+        localityStates.add(localityState);
+        childPickers.put(newLocality, intraLocalitySubchannelStore.currentChildPicker);
+        newState = aggregateState(newState, intraLocalitySubchannelStore.currentChildState);
+      }
+
+      localityStates = Collections.unmodifiableList(localityStates);
+      LocalityPicker localityPicker = new LocalityPicker(localityStates, wrrAlgo);
+      if (!localityPicker.equals(currentPicker.localityPicker)) {
+        childPickers = Collections.unmodifiableMap(childPickers);
+        XdsPicker newXdsPicker = new XdsPicker(localityPicker, childPickers);
```

Comment:
Since only READY localities are considered for picking, you should only pass child pickers with READY state to here. This can be done by `XdslbState` wrapping the helper and intercepting `updateLoadBalancingState()`.

Reply:
Done.